File: src/main/java/mods/eln/sixnode/powersocket/PowerSocketElement.java
package mods.eln.sixnode.powersocket;

import mods.eln.generic.GenericItemUsingDamageDescriptor;
import mods.eln.item.BrushDescriptor;
import mods.eln.item.ConfigCopyToolDescriptor;
import mods.eln.item.IConfigurable;
import mods.eln.misc.Coordinate;
import mods.eln.misc.Direction;
import mods.eln.misc.LRDU;
import mods.eln.misc.Utils;
import mods.eln.node.AutoAcceptInventoryProxy;
import mods.eln.node.NodeBase;
import mods.eln.node.six.SixNode;
import mods.eln.node.six.SixNodeDescriptor;
import mods.eln.node.six.SixNodeElement;
import mods.eln.node.six.SixNodeElementInventory;
import mods.eln.sim.ElectricalLoad;
import mods.eln.sim.IProcess;
import mods.eln.sim.ThermalLoad;
import mods.eln.sim.mna.component.Resistor;
import mods.eln.sim.nbt.NbtElectricalLoad;
import mods.eln.sim.process.destruct.VoltageStateWatchDog;
import mods.eln.sim.process.destruct.WorldExplosion;
import mods.eln.sixnode.electricalcable.ElectricalCableDescriptor;
import mods.eln.sixnode.lampsupply.LampSupplyElement;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.inventory.Container;
import net.minecraft.inventory.IInventory;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.nbt.NBTTagList;
import net.minecraft.nbt.NBTTagString;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.util.List;

public class PowerSocketElement extends SixNodeElement implements IConfigurable {

    //NodeElectricalGateInput inputGate = new NodeElectricalGateInput("inputGate");
    public PowerSocketDescriptor descriptor;

    public NbtElectricalLoad outputLoad = new NbtElectricalLoad("outputLoad");
    public Resistor loadResistor = new Resistor(null, null);  // Connected in process()
    public IProcess PowerSocketSlowProcess = new PowerSocketSlowProcess();

    private AutoAcceptInventoryProxy acceptingInventory = new AutoAcceptInventoryProxy(
        new SixNodeElementInventory(1, 64, this)
    ).acceptIfEmpty(0, ElectricalCableDescriptor.class);

    public String channel = "Default channel";

    public int paintColor = 0;

    VoltageStateWatchDog voltageWatchdog = new VoltageStateWatchDog(outputLoad);

    public static final byte setChannelId = 1;

    @Override
    public IInventory getInventory() {
        return acceptingInventory.getInventory();
    }

    @Nullable
    @Override
    public Container newContainer(@NotNull Direction side, @NotNull EntityPlayer player) {
        return new PowerSocketContainer(player, getInventory());
    }

    public PowerSocketElement(SixNode sixNode, Direction side, SixNodeDescriptor descriptor) {
        super(sixNode, side, descriptor);
        electricalLoadList.add(outputLoad);
        electricalComponentList.add(loadResistor);
        slowProcessList.add(PowerSocketSlowProcess);
        loadResistor.highImpedance();
        this.descriptor = (PowerSocketDescriptor) descriptor;

        slowProcessList.add(voltageWatchdog);
        voltageWatchdog.setDestroys(new WorldExplosion(this).cableExplosion());
    }

    class PowerSocketSlowProcess implements IProcess {

        @Override
        public void process(double time) {
            Coordinate local = sixNode.coordinate;
            LampSupplyElement.PowerSupplyChannelHandle handle = null;
            float bestDist = 1e9f;
            List<LampSupplyElement.PowerSupplyChannelHandle> handles = LampSupplyElement.channelMap.get(channel);
            if(handles != null) {
                for(LampSupplyElement.PowerSupplyChannelHandle hdl : handles) {
                    float dist = (float) hdl.element.sixNode.coordinate.trueDistanceTo(local);
                    if(dist < bestDist && dist <= hdl.element.getRange()) {
                        bestDist = dist;
                        handle = hdl;
                    }
                }
            }

            loadResistor.breakConnection();
            loadResistor.highImpedance();
            if(handle != null && handle.element.getChannelState(handle.id)) {
                ItemStack cable = getInventory().getStackInSlot(PowerSocketContainer.cableSlotId);
                if (cable != null) {
                    ElectricalCableDescriptor desc = (ElectricalCableDescriptor) ElectricalCableDescriptor.getDescriptor(cable);
                    loadResistor.connectTo(handle.element.powerLoad, outputLoad);
                    desc.applyTo(loadResistor);
                }
            }
        }
    }

    @Override
    public ElectricalLoad getElectricalLoad(LRDU lrdu, int mask) {
        if (getInventory().getStackInSlot(PowerSocketContainer.cableSlotId) == null) return null;
        return outputLoad;
    }

    @Nullable
    @Override
    public ThermalLoad getThermalLoad(@NotNull LRDU lrdu, int mask) {
        return null;
    }

    @Override
    public int getConnectionMask(LRDU lrdu) {
        if (getInventory().getStackInSlot(PowerSocketContainer.cableSlotId) == null) return 0;
        return NodeBase.maskElectricalPower + (1 << NodeBase.maskColorCareShift) + (paintColor << NodeBase.maskColorShift);
    }

    @Override
    public String multiMeterString() {
        return Utils.plotUIP(outputLoad.getVoltage(), outputLoad.getCurrent());
    }

    @NotNull
    @Override
    public String thermoMeterString() {
        return null;
    }

    @Override
    public void initialize() {
        setupFromInventory();
    }

    @Override
    public void inventoryChanged() {
        super.inventoryChanged();
        sixNode.disconnect();
        setupFromInventory();
        sixNode.connect();
        needPublish();
    }

    @Override
    public void destroy(EntityPlayerMP entityPlayer) {
        super.destroy(entityPlayer);
    }

    @Override
    public void writeToNBT(NBTTagCompound nbt) {
        super.writeToNBT(nbt);
        nbt.setString("channel", channel);
        nbt.setInteger("color", paintColor);
    }

    @Override
    public void readFromNBT(@NotNull NBTTagCompound nbt) {
        super.readFromNBT(nbt);
        channel = nbt.getString("channel");
        paintColor = nbt.getInteger("color");
    }

    void setupFromInventory() {
        ItemStack cableStack = getInventory().getStackInSlot(PowerSocketContainer.cableSlotId);
        if (cableStack != null) {
            ElectricalCableDescriptor desc = (ElectricalCableDescriptor) ElectricalCableDescriptor.getDescriptor(cableStack);
            desc.applyTo(outputLoad);
            voltageWatchdog.setNominalVoltage(desc.electricalNominalVoltage);
        } else {
            voltageWatchdog.setNominalVoltage(10000);
            outputLoad.highImpedance();
        }
    }

    @Override
    public void networkUnserialize(DataInputStream stream) {
        super.networkUnserialize(stream);

        try {
            switch (stream.readByte()) {
                case setChannelId:
                    channel = stream.readUTF();
                    needPublish();
                    break;
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Override
    public boolean hasGui() {
        return true;
    }

    @Override
    public void networkSerialize(DataOutputStream stream) {
        super.networkSerialize(stream);
        try {
            stream.writeUTF(channel);
            Utils.serialiseItemStack(stream, getInventory().getStackInSlot(PowerSocketContainer.cableSlotId));
            stream.writeInt(paintColor);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Override
    public boolean onBlockActivated(EntityPlayer entityPlayer, Direction side, float vx, float vy, float vz) {
        ItemStack used = entityPlayer.getCurrentEquippedItem();
        if(used != null) {
            GenericItemUsingDamageDescriptor desc = GenericItemUsingDamageDescriptor.getDescriptor(used);
            if(desc != null && desc instanceof BrushDescriptor) {
                BrushDescriptor brush = (BrushDescriptor) desc;
                int color = brush.getColor(used);
                if(color != paintColor && brush.use(used, entityPlayer)) {
                    paintColor = color;
                    sixNode.reconnect();
                }
                return true;
            }
        }

        return acceptingInventory.take(used, this, true, true);
    }

    @Override
    public void readConfigTool(NBTTagCompound compound, EntityPlayer invoker) {
        if(compound.hasKey("powerChannels")) {
            String newChannel = compound.getTagList("powerChannels", 8).getStringTagAt(0);
            if(newChannel != null && newChannel != "") {
                channel = newChannel;
                needPublish();
            }
        }
        if(ConfigCopyToolDescriptor.readCableType(compound, getInventory(), 0, invoker))
            needPublish();
    }

    @Override
    public void writeConfigTool(NBTTagCompound compound, EntityPlayer invoker) {
        NBTTagList list = new NBTTagList();
        list.appendTag(new NBTTagString(channel));
        compound.setTag("powerChannels", list);
        ConfigCopyToolDescriptor.writeCableType(compound, getInventory().getStackInSlot(0));
    }
}

File: src/main/kotlin/mods/eln/sixnode/powersocket/PowerSocketElement.kt
package mods.eln.sixnode.powersocket

import mods.eln.Eln
import mods.eln.generic.GenericItemUsingDamageDescriptor.Companion.getDescriptor
import mods.eln.i18n.I18N
import mods.eln.item.BrushDescriptor
import mods.eln.item.IConfigurable
import mods.eln.misc.*
import mods.eln.misc.Utils.plotPower
import mods.eln.misc.Utils.plotUIP
import mods.eln.misc.Utils.plotVolt
import mods.eln.node.NodeBase
import mods.eln.node.six.SixNode
import mods.eln.node.six.SixNodeDescriptor
import mods.eln.node.six.SixNodeElement
import mods.eln.sim.ElectricalLoad
import mods.eln.sim.IProcess
import mods.eln.sim.ThermalLoad
import mods.eln.sim.mna.component.VoltageSource
import mods.eln.sim.mna.misc.MnaConst
import mods.eln.sim.nbt.NbtElectricalLoad
import mods.eln.sim.process.destruct.VoltageStateWatchDog
import mods.eln.sim.process.destruct.WorldExplosion
import mods.eln.sixnode.lampsupply.AvailableSupply
import mods.eln.sixnode.lampsupply.IWirelessPower
import mods.eln.sixnode.lampsupply.LampSupplyConnectionHelper
import mods.eln.sixnode.lampsupply.PowerChannelTextboxHelper
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.nbt.NBTTagCompound
import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.IOException
import kotlin.math.abs
import kotlin.math.pow

class PowerSocketElement(sixNode: SixNode?, side: Direction?, descriptor: SixNodeDescriptor?) : SixNodeElement(
    sixNode!!, side!!, descriptor!!
), IConfigurable {
    var descriptor: PowerSocketDescriptor?
    var electricalLoad = NbtElectricalLoad("electricalLoad")
    var voltageSource = VoltageSource("voltSrc", electricalLoad, null)
    private var powerSocketProcess: IProcess = PowerSocketProcess(this)
    var channel = PowerChannelTextboxHelper.DEFAULT_CHANNEL_STRING
    var activeLampSupplyConnection = false
    private var paintColor = 0
    var voltageWatchdog = VoltageStateWatchDog(electricalLoad)

    init {
        electricalLoadList.add(electricalLoad)
        electricalComponentList.add(voltageSource)
        slowProcessList.add(powerSocketProcess)
        this.descriptor = descriptor as PowerSocketDescriptor?
        slowProcessList.add(voltageWatchdog)
        voltageWatchdog.setDestroys(WorldExplosion(this).cableExplosion())
        voltageWatchdog.setNominalVoltage(NominalVoltage.V240)
    }

    class PowerSocketProcess(val element: PowerSocketElement) : IProcess, IWirelessPower {
        override var previousConnectedSupply: AvailableSupply? = null

        override val powerChannel: String
            get() = element.channel
        override val coordinate: Coordinate
            get() = element.coordinate!!
        override val loadResistance: Double
            get() {
                return if (abs(element.voltageSource.power) < MnaConst.almostZero) MnaConst.highImpedance
                else element.voltageSource.voltage.pow(2) / abs(element.voltageSource.power)
            }

        override fun updateLoadVoltage(newVoltage: Double) {
            element.voltageSource.voltage = newVoltage
        }

        override fun process(time: Double) {
            element.activeLampSupplyConnection = LampSupplyConnectionHelper.connectToLampSupply(this)
            element.needPublish()
        }
    }

    override fun getElectricalLoad(lrdu: LRDU, mask: Int): ElectricalLoad {
        return electricalLoad
    }

    override fun getThermalLoad(lrdu: LRDU, mask: Int): ThermalLoad? {
        return null
    }

    override fun getConnectionMask(lrdu: LRDU): Int {
        return NodeBase.maskElectricalPower + (1 shl NodeBase.maskColorCareShift) + (paintColor shl NodeBase.maskColorShift)
    }

    override fun multiMeterString(): String {
        return plotUIP(electricalLoad.voltage, electricalLoad.getCurrent())
    }

    override fun thermoMeterString(): String {
        return ""
    }

    override fun getWaila(): Map<String, String> {
        val info: MutableMap<String, String> = LinkedHashMap()

        info[I18N.tr("Power Provided")] = plotPower("", voltageSource.power)

        if (Utils.isWailaEasyModeEnabled()) {
            info[I18N.tr("Voltage")] = plotVolt("", voltageSource.voltage)
            info[I18N.tr("Channel")] = channel
        }

        return info
    }

    override fun initialize() {
        Eln.applySmallRs(electricalLoad)
    }

    override fun inventoryChanged() {
        super.inventoryChanged()
        sixNode!!.disconnect()
        sixNode!!.connect()
        needPublish()
    }

    override fun writeToNBT(nbt: NBTTagCompound) {
        super.writeToNBT(nbt)
        nbt.setString("channel", channel)
        nbt.setInteger("color", paintColor)
    }

    override fun readFromNBT(nbt: NBTTagCompound) {
        super.readFromNBT(nbt)
        channel = nbt.getString("channel")
        paintColor = nbt.getInteger("color")
    }

    override fun networkUnserialize(stream: DataInputStream) {
        super.networkUnserialize(stream)
        try {
            when (stream.readByte()) {
                SET_CHANNEL_EVENT -> {
                    channel = stream.readUTF()
                    needPublish()
                }
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    override fun hasGui(): Boolean {
        return true
    }

    override fun networkSerialize(stream: DataOutputStream) {
        super.networkSerialize(stream)
        try {
            stream.writeUTF(channel)
            stream.writeBoolean(activeLampSupplyConnection)
            stream.writeInt(paintColor)
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    override fun onBlockActivated(
        entityPlayer: EntityPlayer,
        side: Direction,
        vx: Float,
        vy: Float,
        vz: Float
    ): Boolean {
        val used = entityPlayer.currentEquippedItem
        if (used != null) {
            val desc = getDescriptor(used)
            if (desc != null && desc is BrushDescriptor) {
                val color: Int = desc.getColor(used)
                if (color != paintColor && desc.use(used, entityPlayer)) {
                    paintColor = color
                    sixNode!!.reconnect()
                }
                return true
            }
        }
        return false
    }

    override fun readConfigTool(compound: NBTTagCompound, invoker: EntityPlayer) {
        if (compound.hasKey("lampSupplyChannel")) {
            channel = compound.getString("lampSupplyChannel")
            needPublish()
        }
    }

    override fun writeConfigTool(compound: NBTTagCompound, invoker: EntityPlayer) {
        compound.setString("lampSupplyChannel", channel)
    }

    companion object {
        const val SET_CHANNEL_EVENT: Byte = 1
    }
}
